File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3067_count-pairs-of-connectable-servers-in-a-weighted-tree-network/solutions/java/solution.java
import java.util.*;

class Solution {
    private ArrayList<int[]>[] graph;
    private HashMap<Integer, Integer>[] distanceCount;
    private int[] result;
    
    public int[] countPairsOfConnectableServers(int[][] edges, int signalSpeed) {
        int n = edges.length + 1;
        graph = new ArrayList[n];
        result = new int[n];
        distanceCount = new HashMap[n];
        
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
            distanceCount[i] = new HashMap<>();
        }
        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int weight = edge[2];
            graph[u].add(new int[]{v, weight});
            graph[v].add(new int[]{u, weight});
        }
        
        for (int i = 0; i < n; i++) {
            dfs(i, -1, 0, distanceCount[i]);
            calculatePairs(i, signalSpeed, distanceCount[i]);
        }
        
        return result;
    }
    
    private void dfs(int current, int parent, int currentDist, HashMap<Integer, Integer> map) {
        map.put(currentDist, map.getOrDefault(currentDist, 0) + 1);
        
        for (int[] adj : graph[current]) {
            int neighbor = adj[0];
            int weight = adj[1];
            if (neighbor != parent) {
                dfs(neighbor, current, currentDist + weight, map);
            }
        }
    }
    
    private void calculatePairs(int node, int signalSpeed, HashMap<Integer, Integer> map) {
        int count = 0;
        
        for (int dist : map.keySet()) {
            if (dist % signalSpeed == 0) {
                int numServers = map.get(dist);
                count += (numServers * (numServers - 1)) / 2;
            }
        }
        
        result[node] = count;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3067_count-pairs-of-connectable-servers-in-a-weighted-tree-network/solutions/python3/solution.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def countPairsOfConnectableServers(self, edges: List[List[int]], signalSpeed: int) -> List[int]:
        n = len(edges) + 1
        
        # Adjacency list for the tree
        graph = defaultdict(list)
        for a, b, weight in edges:
            graph[a].append((b, weight))
            graph[b].append((a, weight))
        
        # Function to calculate distances from a given node
        def calculate_distances(start):
            dist = [-1] * n
            dist[start] = 0
            queue = deque([start])
            while queue:
                node = queue.popleft()
                current_distance = dist[node]
                for neighbor, weight in graph[node]:
                    if dist[neighbor] == -1:  # Unvisited
                        dist[neighbor] = current_distance + weight
                        queue.append(neighbor)
            return dist
        
        # Count pairs of nodes (a, b) such that both distances are divisible by signalSpeed
        # and there are no shared edges in their paths to the node `c`
        count = [0] * n
        
        # Precompute distances from all nodes
        distances = [calculate_distances(i) for i in range(n)]
        
        # Check each node if it can be a central node `c`
        for c in range(n):
            # Create a mapping of distances modulus signalSpeed
            mod_map = defaultdict(list)
            for i in range(n):
                if i != c:
                    mod_map[distances[c][i] % signalSpeed].append(i)
            
            # For each list of nodes at the same modulo class, calculate pairs
            for nodes in mod_map.values():
                if len(nodes) >= 2:
                    # Convert node list to index based on distance, and count pairs
                    sorted_nodes = sorted(nodes, key=lambda x: distances[c][x])
                    size = len(sorted_nodes)
                    for i in range(size):
                        for j in range(i + 1, size):
                            a, b = sorted_nodes[i], sorted_nodes[j]
                            if a < b and distances[c][a] != -1 and distances[c][b] != -1:
                                if (distances[c][b] - distances[c][a]) % signalSpeed == 0:
                                    count[c] += 1

        return count
